### ISSUE TITLE
Unable to use workflowTemplateRef with defaults workflow serviceAccountName

### ISSUE THREAD

**Issue by stefansedich** [argoproj/argo#3355]
Checklist:

 * [X] I've included the version.
 * [X] I've included reproduction steps.
 * [X] I've included the workflow YAML.
 * [ ] I've included the logs.
 
**What happened**:

Ever since the validation changes were introduced in https://github.com/argoproj/argo/pull/2912 we are unable to submit a workflow using a `workflowTemplateRef` when there is an `serviceAccountName` defined in workflow defaults, given the following defaults:

```
workflowDefaults:
  serviceAccountName: argo-workflow
```

When submitting the following workflow:

```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  name: foo-fw4vp
  namespace: models
spec:
  arguments:
    parameters:
  serviceAccountName: argo-workflow
  workflowTemplateRef:
    name: foo
```

 it raises the following error:

```ServiceAccountName is invalid field in spec if workflow referred WorkflowTemplate reference```

**What you expected to happen**:

That this will work as it did prior to these changes.

**How to reproduce it (as minimally and precisely as possible)**:

Configure argo with the workflow defaults configmap defining a default `serviceAccountName` and submit a workflow using a `workflowTemplateRef`.

**Environment**:

- Argo version: `2.9.0-RC4`
- Kubernetes version : `1.15`

---

**Message from the maintainers**:

If you are impacted by this bug please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

**Comment by stefansedich** [comment]
@sarabala1979 here is the issue as discussed on Slack.

**Comment by sarabala1979** [comment]
I will work on it

**Comment by vladlosev** [comment]
Mmm, we use some other workflow spec fields in the workflow defaults section of the config map. These include `nodeSelector` and `tolerations` in order to restrict our workflow pods to spot instances, and `ttlStrategy` to set the default TTL on all the workflows.

**Comment by alexec** [comment]
@sarabala1979 I think the fix is this:

```
diff --git a/workflow/controller/operator.go b/workflow/controller/operator.go
index cbd818b9..b4c8e7f3 100644
--- a/workflow/controller/operator.go
+++ b/workflow/controller/operator.go
@@ -2816,7 +2816,12 @@ func (woc *wfOperationCtx) loadExecutionSpec() (wfv1.TemplateReferenceHolder, wf
                }
        }
 
-       woc.wfSpec = woc.wf.Status.StoredWorkflowSpec
+       storedWorkflow := &wfv1.Workflow{Spec: *woc.wf.Status.StoredWorkflowSpec}
+       err := woc.controller.setWorkflowDefaults(storedWorkflow)
+       if err != nil {
+               return nil, executionParameters, err
+       }
+       woc.wfSpec = &storedWorkflow.Spec
 
        entrypoint := woc.wf.Spec.Entrypoint
        if entrypoint == "" {
```

**Comment by sarabala1979** [comment]
> @sarabala1979 I think the fix is this:
> 
> ```
> diff --git a/workflow/controller/operator.go b/workflow/controller/operator.go
> index cbd818b9..b4c8e7f3 100644
> --- a/workflow/controller/operator.go
> +++ b/workflow/controller/operator.go
> @@ -2816,7 +2816,12 @@ func (woc *wfOperationCtx) loadExecutionSpec() (wfv1.TemplateReferenceHolder, wf
>                 }
>         }
>  
> -       woc.wfSpec = woc.wf.Status.StoredWorkflowSpec
> +       storedWorkflow := &wfv1.Workflow{Spec: *woc.wf.Status.StoredWorkflowSpec}
> +       err := woc.controller.setWorkflowDefaults(storedWorkflow)
> +       if err != nil {
> +               return nil, executionParameters, err
> +       }
> +       woc.wfSpec = &storedWorkflow.Spec
>  
>         entrypoint := woc.wf.Spec.Entrypoint
>         if entrypoint == "" {
> ```

I thought about this. This will resolve the only defaults. I am thinking  more usecases if user passing serviceaccount through CLI option, need to pass volumes and pvc.

**Comment by vladlosev** [comment]
Validation failures are still happening after applying #3480 when I put tolerations into `workflowDefaults`:
> Failed: invalid spec: Tolerations is invalid field in spec if workflow referred WorkflowTemplate reference.

**Comment by alexec** [comment]
Fixed and available in v2.10.0-rc5.
